### ISSUE TITLE
Crash in taskIdentifier.get

### ISSUE THREAD

**Issue by hijamoya** [onevcat/Kingfisher#1603]
### Check List

Thanks for considering to open an issue. Before you submit your issue, please confirm these boxes are checked.

- [x] I have read the [wiki page](https://github.com/onevcat/Kingfisher/wiki) and [cheat sheet](https://github.com/onevcat/Kingfisher/wiki/Cheat-Sheet), but there is no information I need.
- [x] I have searched in [existing issues](https://github.com/onevcat/Kingfisher/issues?utf8=✓&q=is%3Aissue), but did not find a same one.
- [x] I want to report a problem instead of asking a question. It'd better to use [kingfisher tag in Stack Overflow](http://stackoverflow.com/questions/tagged/kingfisher) to ask a question.

### Issue Description

We receive a crash after upgrade to 6.0.1, this never happened before 6.0.0 version.

Here is the trace:

```
Hardware Model:     iPhone9,2
Process:            shopping
Identifier:         com.lativ.iShopping.CN
Version:            2.2.4
Role:               Foreground
OS Version:         iOS 11.2.6
Exception Type:     EXC_BAD_ACCESS 


EXC_BAD_ACCESS: Attempted to dereference garbage pointer 0xa6d09bec8.

0  libobjc.A.dylib         _objc_retain
1  Kingfisher              merged generic specialization <Kingfisher.Box<Kingfisher.Indicator>> of Kingfisher.getAssociatedObject<A>(Any, Swift.UnsafeRawPointer) -> A?
2  shopping                taskIdentifier.get (ImageView+Kingfisher.swift:436:54)
3  shopping                onDataReceived (KingfisherOptionsInfo.swift:382:20)
4  shopping                Kingfisher.SessionDelegate.urlSession(_: __C.NSURLSession, dataTask: __C.NSURLSessionDataTask, didReceive: Foundation.Data) -> () (SessionDelegate.swift:184:28)
5  <compiler-generated>    urlSession
6  CFNetwork               ___67-[NSURLSession delegate_dataTask:didReceiveData:completionHandler:]_block_invoke.264
7  Foundation              ___NSBLOCKOPERATION_IS_CALLING_OUT_TO_A_BLOCK__
8  Foundation              -[NSBlockOperation main]
9  Foundation              -[__NSOperationInternal _start:]
10 libdispatch.dylib       __dispatch_client_callout
11 libdispatch.dylib       __dispatch_block_invoke_direct$VARIANT$mp
12 libdispatch.dylib       __dispatch_client_callout
13 libdispatch.dylib       __dispatch_block_invoke_direct$VARIANT$mp
14 libdispatch.dylib       _dispatch_block_perform$VARIANT$mp
15 Foundation              ___NSOQSchedule_f
16 libdispatch.dylib       __dispatch_client_callout
17 libdispatch.dylib       __dispatch_continuation_pop$VARIANT$mp
18 libdispatch.dylib       __dispatch_async_redirect_invoke$VARIANT$mp
19 libdispatch.dylib       __dispatch_root_queue_drain
20 libdispatch.dylib       __dispatch_worker_thread3
21 libsystem_pthread.dylib __pthread_wqthread
```

**Comment by onevcat** [comment]
There was no change recently on this. Not sure what is going on. 

> this never happened before 6.0.0 version.

Is that containing 6.0.0 or only versions 5.x?

I'd wait for more information/report or reproducible steps before we can dig into and fix it. :(

**Comment by hijamoya** [comment]
@onevcat We just upgrade from 5.0 version to 6.0.1 this week, the app have launched for 3 months and never happened this crash before.

**Comment by dominiquemiller** [comment]
@onevcat @hijamoya  I can confirm this error as well. Last week I updated from Kingfisher 5.14 to 6.1.0.  We are using the Swift package manager.  100% SwiftUI App.  We began seeing this error this Monday during our internal testing.  Happens in the KFImage file line 173.
![KFImage swift 2021-02-09 14-41-27](https://user-images.githubusercontent.com/14062553/107419386-be700700-6ae5-11eb-92bd-ed9738ea19c6.png)

**Comment by dominiquemiller** [comment]
@onevcat @hijamoya FYI when I revert back to version 6.0.1 the error does not appear.

**Comment by onevcat** [comment]
@dominiquemiller 
Thanks for reporting this.

Do you have a way to reproduce it in a predictable way?
The crash happened in an unbelievable location. Is it possible for you to also paste the full stack (the truncated part of the second image)?

**Comment by onevcat** [comment]
6.0.1 does not use this implementation. The local state setting is from 6.1.0. So I guess it is some internal SwiftUI thing, which failed when managing the view state in this way. But I myself have never seen this crash so it is hard to say.

I need a sample or a way to reproduce it first.

**Comment by dominiquemiller** [comment]
> @dominiquemiller
> Thanks for reporting this.
> 
> Do you have a way to reproduce it in a predictable way?
> The crash happened in an unbelievable location. Is it possible for you to also paste the full stack (the truncated part of the second image)?

Here is the whole stack trace from TestFlight:
![crashlog crash 2021-02-10 09-05-06](https://user-images.githubusercontent.com/14062553/107520162-1d358f00-6b7f-11eb-8d09-900072b0b268.png)

This is the how we are using Kingfisher:

![CompanyLogoView swift 2021-02-10 09-07-01](https://user-images.githubusercontent.com/14062553/107520498-7a314500-6b7f-11eb-8be9-a882beff60f4.png)

I ended up reverting back to 6.0.1 and so far the error has not happened.  Its very weird so I agree with you that it's something internal with SwiftUI.

**Comment by dominiquemiller** [comment]
@onevcat Going thru the other crash reports this one is better formatted.
![EXC_BAD_ACCESS: Exception 1, Code 0, Subcode 8   Attempted to dereference garbage pointer 0xbf9e00000000  2021-02-11 09-22-55](https://user-images.githubusercontent.com/14062553/107649056-0bb9b900-6c4b-11eb-97f3-66de6b13c69e.png)

**Comment by onevcat** [comment]
@dominiquemiller 

Thanks for the detailed stack. It indeed relates to some state management. Although the root reason for this is still mysterious, there are already quite a few critical fixes around the internal state management and maybe they can fix it.

I think currently the new state implementation introduced from 6.1.0 is quite stable on the master branch. If you'd like, it worths trying the master branch and see how it works. I will keep an eye on this and also see if I can reproduce it locally!

**Comment by rrozenv** [comment]
I am also seeing this exact issue on 6.1.0 and 6.1.1. Reverting to 6.0.1 seems to be working though

**Comment by windom** [comment]
We've experienced this also; upgrading from 6.0.1 to 6.1.1 resulted in tons of crashes like the one posted by @dominiquemiller. Downgrading back made the crashes disappear. Trying to find a repro.

**Comment by onevcat** [comment]
@windom Thanks for the reporting! I am also digging into it (but without luck of reproducing it in any of my samples). Please let me know if you can reproduce it in any way!

**Comment by onevcat** [comment]
I guess I found a potential problem with it and here is a fix in the branch `fix/unexpected-ref`. Since I cannot reproduce this crash myself, I need your help to try if it is a valid fix. It would be a great help if you can try that branch and let me know how it is working. If everything goes fine, I will merge it and tag a new release.

**Comment by windom** [comment]
For me the code below on 14.4, SE2, after a bit of scrolling reproduces the crash reliably on 6.1.1.
On 6.0.1 and `fix/unexpected-ref` the crash does not happen 👍 

```import SwiftUI
import Kingfisher

let urls = [
    "https://www.amazon.com/favicon.ico",
    "https://s.ytimg.com/yts/img/favicon_96-vflW9Ec0w.png",
    "https://www.wikipedia.org/static/apple-touch/wikipedia.png",
    "https://github.githubassets.com/favicons/favicon.svg",
    "https://github.githubassets.com/apple-touch-icon-180x180.png",
]

struct ContentView: View {
    var body: some View {
        List {
            ForEach(1..<200) { id in
                KFImage(URL(string: urls[id%urls.count]))
                    .renderingMode(.original)
                    .resizable()
                    .frame(width: 24, height: 24)
            }
        }
    }
}
````

**Comment by onevcat** [comment]
I am closing this since the fixes should be contained in 6.2.0 and I am going to tag it now.
